### PR TITLE
random_pet: Random pet name generator

### PR DESCRIPTION
**Pull request by Akasurde**
##### SUMMARY

A lookup plugin to generate random pet names based
upon criteria.

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
plugins/lookup/random_pet.py
tests/integration/targets/lookup_random_pet/aliases
tests/integration/targets/lookup_random_pet/dependencies.yml
tests/integration/targets/lookup_random_pet/runme.sh
tests/integration/targets/lookup_random_pet/test.yml

### CODE REVIEW

File: plugins/lookup/random_pet.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+from __future__ import (absolute_import, division, print_function)
+
+__metaclass__ = type
+
+DOCUMENTATION = r'''
+    name: random_pet
+    author:
+      - Abhijeet Kasurde (@Akasurde)
+    short_description: Generates random pet names
+    version_added: '3.0.2'
```
Comment (felixfontein) [inline plugins/lookup/random_pet.py:15]:
```suggestion
    version_added: '3.1.0'
```

File: plugins/lookup/random_pet.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    requirements:
+      - Python-petname U(https://github.com/dustinkirkland/python-petname).
```
Comment (felixfontein) [inline plugins/lookup/random_pet.py:17]:
```suggestion
      - petname U(https://github.com/dustinkirkland/python-petname).
```
After all it's `pip install petname`.

File: plugins/lookup/random_pet.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    description:
+      - Generates random pet names that can be used as unique identifiers for the resources.
+    options:
+      words:
+        description:
+        - The number of words in the pet name.
+        default: 2
```
Comment (felixfontein) [inline plugins/lookup/random_pet.py:24]:
```suggestion
        default: 2
        type: int
```

File: plugins/lookup/random_pet.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+      length:
+        description: The length of the pet name.
+        default: 6
```
Comment (felixfontein) [inline plugins/lookup/random_pet.py:27]:
```suggestion
        default: 6
        type: int
```

File: plugins/lookup/random_pet.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+- name: Generate pet name with length
+  ansible.builtin.debug:
+    var: lookup('community.general.random_pet', length=7)
+'''
+
+RETURN = r'''
+  _raw:
+    description: Random pet name
+    type: str
```
Comment (felixfontein) [inline plugins/lookup/random_pet.py:56]:
```suggestion
    type: list
    elements: str
```
The plugin returns a list of strings (with exactly one string), not a string.

File: plugins/lookup/random_pet.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    var: lookup('community.general.random_pet', separator="_")
+
+- name: Generate pet name with length
+  ansible.builtin.debug:
+    var: lookup('community.general.random_pet', length=7)
+'''
+
+RETURN = r'''
+  _raw:
+    description: Random pet name
```
Comment (felixfontein) [inline plugins/lookup/random_pet.py:55]:
```suggestion
    description: A one-element list containing a random pet name.
```

File: plugins/lookup/random_pet.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+class LookupModule(LookupBase):
+
+    def run(self, terms, variables=None, **kwargs):
+
+        if not HAS_PETNAME:
+            raise AnsibleError('Python petname library is required. '
+                               'Please install using "pip install petname"')
+
+        words = kwargs.get('words', 2)
```
Comment (felixfontein) [inline plugins/lookup/random_pet.py:78]: Please use proper option handling with `set_options()`/`get_option()`.

File: plugins/lookup/random_pet.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+      separator:
+        description: The character to separate words in the pet name.
+        default: "-"
+        type: str
+'''
+
+EXAMPLES = r'''
+- name: Generate pet name
+  ansible.builtin.debug:
+    var: lookup('community.general.random_pet')
```
Comment (felixfontein) [inline plugins/lookup/random_pet.py:42]: How about adding an example output next to every example, like:
```suggestion
    var: lookup('community.general.random_pet')
    # Example result: 'lasting-toucan'
```
Same for the others. I think that makes it easier to understand what these options do, and what the lookup does in general.

**Comment by aminvakil** [comment]
Is this really fit to include in `community.general`? :)

File: tests/integration/targets/lookup_random_pet/test.yml [hunk 9]
```diff
@@ -0,0 +1,10 @@
+- hosts: localhost
```
Comment (felixfontein) [inline tests/integration/targets/lookup_random_pet/test.yml:1]:
```suggestion
- hosts: localhost
  gather_facts: false
```

File: tests/integration/targets/lookup_random_pet/test.yml [hunk 10]
```diff
@@ -0,0 +1,10 @@
...
+  tasks:
+
+  - assert:
+      that:
+      - query('community.general.random_pet', words=3) | length == 1
+      - query('community.general.random_pet', length=3) | length == 1
+      - query('community.general.random_pet', prefix="kuberentes") | length == 1
+      - query('community.general.random_pet', separator="_") | length == 1
+      - query('community.general.random_pet', words=2, length=6, prefix="kuberentes", separator="_") | length == 1
```
Comment (felixfontein) [inline tests/integration/targets/lookup_random_pet/test.yml:10]:
```suggestion
  - name: Call plugin
    set_fact:
      result1: '{{ query('community.general.random_pet', words=3) }}'
      result2: '{{ query('community.general.random_pet', length=3) }}'
      result3: '{{ query('community.general.random_pet', prefix="kuberentes") }}'
      result4: '{{ query('community.general.random_pet', separator="_") }}'
      result5: '{{ query('community.general.random_pet', words=2, length=6, prefix="kuberentes", separator="_") }}'

  - name: Check results
    assert:
      that:
        - result1 | length == 1
        - result2 | length == 1
        - result3 | length == 1
        - result4 | length == 1
        - result5 | length == 1
```
Now it's easier to add more tests, like checking whether `result1[0]` actually has three words etc.

File: plugins/lookup/random_pet.py [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+    description:
+      - Generates random pet names that can be used as unique identifiers for the resources.
+    options:
+      words:
+        description:
+        - The number of words in the pet name.
+        default: 2
+        type: int
+      length:
+        description: The length of the pet name.
```
Comment (felixfontein) [inline plugins/lookup/random_pet.py:27]:
```suggestion
        description:
        - The maximal length of every component of the pet name.
        - Values below 3 will be set to 3 by petname.
```

File: tests/integration/targets/lookup_random_pet/test.yml [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+      result4: "{{ query('community.general.random_pet', separator='_') }}"
+      result5: "{{ query('community.general.random_pet', words=2, length=6, prefix='kubernetes', separator='_') }}"
+
+  - name: Check results
+    assert:
+      that:
+        - result1 | length == 1
+        - result1[0].split('-') | length == 3
+        - result2 | length == 1
+        - result2[0].split('-')[0] | length >= 3
```
Comment (felixfontein) [inline tests/integration/targets/lookup_random_pet/test.yml:18]:
```suggestion
        - result2[0].split('-')[0] | length <= 3
```

**Review by felixfontein — COMMENTED** [review]
This currently fails because `result2` was `on-pig`. I checked the code of petname to see what is happening, and it turns out that values below 3 are set to 3 (https://github.com/dustinkirkland/python-petname/blob/master/petname/__init__.py#L50-L51) and this is only an upper bound on the part length (https://github.com/dustinkirkland/python-petname/blob/master/petname/__init__.py#L31).

**Review by felixfontein — APPROVED** [review]

**Comment by felixfontein** [comment]
I'll merge in a couple of days if nobody complains.

**Comment by felixfontein** [comment]
@Akasurde thanks for contributing this!
@aminvakil thanks for reviewing!

**Pull request merged**